pkg/compactor/block_upload_test.go

```diff
@@ -1133,102 +1136,109 @@ func TestMultitenantCompactor_HandleBlockUpload_Complete(t *testing.T) {
-				bkt.MockExists(metaPath, true, nil)
+			setUpBucket: func(t *testing.T, bkt objstore.Bucket) {
+				err := marshalAndUploadToBucket(context.Background(), bkt, metaPath, validMeta)
+				require.NoError(t, err)
 			},
 			expConflict: "block already exists",
 		},
 		{
-			name:     "checking for complete block fails",
-			tenantID: tenantID,
-			blockID:  blockID,
-			setUpBucketMock: func(bkt *bucket.ClientMock) {
-				bkt.MockExists(metaPath, false, fmt.Errorf("test"))
-			},
+			name:                   "checking for complete block fails",
+			tenantID:               tenantID,
+			blockID:                blockID,
+			errorInjector:          bucket.InjectErrorOn(bucket.OpExists, metaPath, injectedError),
 			expInternalServerError: true,
 		},
 		{
-			name:     "missing in-flight meta file",
-			tenantID: tenantID,
-			blockID:  blockID,
-			setUpBucketMock: func(bkt *bucket.ClientMock) {
-				bkt.MockExists(metaPath, false, nil)
-				setUpGet(bkt, uploadingMetaPath, nil, bucket.ErrObjectDoesNotExist)
-			},
+			name:        "missing in-flight meta file",
+			tenantID:    tenantID,
+			blockID:     blockID,
 			expNotFound: "block upload not started",
 		},
 		{
 			name:     "downloading in-flight meta file fails",
 			tenantID: tenantID,
 			blockID:  blockID,
-			setUpBucketMock: func(bkt *bucket.ClientMock) {
-				bkt.MockExists(metaPath, false, nil)
-				setUpGet(bkt, uploadingMetaPath, nil, fmt.Errorf("test"))
+			setUpBucket: func(t *testing.T, bkt objstore.Bucket) {
+				err := marshalAndUploadToBucket(context.Background(), bkt, uploadingMetaPath, validMeta)
+				require.NoError(t, err)
 			},
+			errorInjector:          bucket.InjectErrorOn(bucket.OpGet, uploadingMetaPath, injectedError),
 			expInternalServerError: true,
 		},
 		{
 			name:     "corrupt in-flight meta file",
 			tenantID: tenantID,
 			blockID:  blockID,
-			setUpBucketMock: func(bkt *bucket.ClientMock) {
-				bkt.MockExists(metaPath, false, nil)
-				setUpGet(bkt, uploadingMetaPath, []byte("{"), nil)
+			setUpBucket: func(t *testing.T, bkt objstore.Bucket) {
+				err := bkt.Upload(context.Background(), uploadingMetaPath, bytes.NewReader([]byte("{")))
+				require.NoError(t, err)
 			},
 			expInternalServerError: true,
 		},
 		{
-			name:     "uploading meta file fails",
-			tenantID: tenantID,
-			blockID:  blockID,
-			setUpBucketMock: func(bkt *bucket.ClientMock) {
-				bkt.MockExists(metaPath, false, nil)
-				metaJSON, err := json.Marshal(validMeta)
-				require.NoError(t, err)
-				setUpGet(bkt, uploadingMetaPath, metaJSON, nil)
-				setUpGet(bkt, validationPath, nil, bucket.ErrObjectDoesNotExist)
-				bkt.MockUpload(metaPath, fmt.Errorf("test"))
+			name:          "uploading meta file fails",
+			tenantID:      tenantID,
+			blockID:       blockID,
+			setUpBucket:   validSetup,
+			errorInjector: bucket.InjectErrorOn(bucket.OpUpload, metaPath, injectedError),
+			expUploadStateResult: &BlockUploadStateResult{
+				State: "failed",
```

Comment:
Nit: We may want to move these state strings into constants.

Reply:
This usage got removed. Right now the states are constants, but they're an `iota` that get switch-cased into the string. I don't feel strongly one way or the other. This did remind me we should add a state for a deletion marker being present though.